src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -1438,12 +1438,15 @@ class RsFnInferenceContext(
     }
 
     private fun RsCondition.inferTypes() {
-        val pat = pat
-        if (pat != null) {
+        val patList = patList
+        if (patList.isNotEmpty()) {
             // if let Some(a) = ... {}
+            // if let V1(a) | V2(a) = ... {}
             // or
             // while let Some(a) = ... {}
-            pat.extractBindings(resolveTypeVarsWithObligations(expr.inferType()))
+            // while let V1(a) | V2(a) = ... {}
+            val exprTy = expr.inferType()
+            patList.forEach { it.extractBindings(resolveTypeVarsWithObligations(exprTy)) }
```

Comment:
The same logic is used inside `inferMatchExprType`. So IMO it should be extracted like this (also `resolveTypeVarsWithObligations(exprTy)` will be computed once):
```Kotlin
private fun RsOrPats.inferTypes(matchingExprTy: Ty) {
    patList.forEach { it.extractBindings(matchingExprTy) }
}

private fun inferMatchExprType(expr: RsMatchExpr, expected: Ty?): Ty {
    // ...
        arm.orPats.inferTypes(matchingExprTy)
    // ...
}

private fun RsCondition.inferTypes() {
    val orPats = orPats
    if (orPats != null) {
        val matchingExprTy = resolveTypeVarsWithObligations(expr.inferType())
        orPats.inferTypes(matchingExprTy)
    } else {
        expr.inferType(TyBool)
    }
}
```